### DISCUSSION TITLE
Collapsible groups.

### DISCUSSION THREAD

**Discussion by Asagrim** [zombieFox/nightTab#277] [Ideas]
Collapsible groups would:
- alleviate size limitation issues on smaller/shallower (ultrawide) screens which restricts the number of possible bookmarks shown on the screen
- would task the system less if someone makes heavy use of animated bookmark backgrounds,
   as they would only be loaded on-demand when collapsed groups are expanded/"uncollapsed" (I already start having playback frame rate/frame drop issues with five looped webm backgrounds each less than 10MB in size with a 2700X,16GB RAM, Gen3 M.2, 500/200 fiber, and sometimes the playback just stops).
- could address privacy issues in a public setting (portable computers and live streaming where browser is being seen by more people than just its user)

How they should optimally work:
- all groups are expanded by default
- an up/down arrow in style of the 'open all' button would be placed after the group name
- collapsed/uncollapsed state would be remembered by the extension (much like edit mode, when I close a new tab with edit mode on, another one opens later with it on)
- any bookmark background (image/video) would be loaded/unloaded upon uncollapsing/collapsing a group, to aid system stability

**Comment by zombieFox** [comment]
Hi, thanks for adding this. I can see the benefit of such a feature. This has been mentioned in a couple of issues (e.g. #237). 

I did develop a test branch for this some time ago. One thing I confirmed, which is worth pointing out, is animating CSS height is traditionally very hard. So hiding or showing the bookmark groups will very likely utilise `display: none`, (I would prefer to animate the height to 0 and back again).

This will be out of step with most of the UI in nightTab which employ animations and transitions to help describe to users what is happening on screen.

**Reply by Asagrim** [reply]
Thank you for your response, I understand what you mean. The way I see it as an end-user, it would be a very minor trade-off for me. Currently I can have five units of bookmark height on my screen before the rest is starting to go off-screen, and that's with the header already completely disabled. If I make the bookmarks any smaller, their image logos would cover too much of their background image if I want to keep them legible, so one would defeat the purpose of the other.

With the whole point of the extension being geometrically adjustable bookmarks, that doesn't leave me with much choice for layouts, if I want to have several groups. Basically I can have one tall tile in one group, and a plain list of 1x1 tiles in the other groups. Collapsible groups would make this issue disappear, along with the remaining issues I mentioned, the second worst of which being the performance issues creeping on using "too many" animations.

With how many issues it would solve regarding usability, a transition effect for eye-candy becomes irrelevant.

**Reply by zombieFox** [reply]
I agree, the trade off is certainly worth the functionality. Just setting some expectations. Also, I've got a prototype in the works: 

![Kapture 2021-09-08 at 09 06 27](https://user-images.githubusercontent.com/7115017/132471100-1d37875f-afe1-4c33-a94c-dcf4d714818b.gif)

**Reply by zombieFox** [reply]
> ... I already start having playback frame rate/frame drop issues with five looped webm backgrounds each less than 10MB in size with a 2700X,16GB RAM, Gen3 M.2, 500/200 fiber, and sometimes the playback just stops...

Any chance I can get your backup JSON so I can investigate these performance issues?

**Reply by Asagrim** [reply]
> > ... I already start having playback frame rate/frame drop issues with five looped webm backgrounds each less than 10MB in size with a 2700X,16GB RAM, Gen3 M.2, 500/200 fiber, and sometimes the playback just stops...
> 
> Any chance I can get your backup JSON so I can investigate these performance issues?

That gif is the exact representation of what I imagined when I wrote the feature request, nice! :D

As for the performance issue:

I didn't want to waste your time sending you on a wild goose chase, so I took it upon myself to diagnose the issue I'm having. Turns out it's connected to the length of the browser session I have had running, so a JSON file in itself that shows no performance issue when loaded with less than a dozen open tabs in a freshly started browser won't be useful in diagnosing much of anything. What I don't understand is why it's only the new tab that starts having performance issues with a browser that ran for a long time. Every other tab I have already open or will open is completely fine even if it has been 3 days since I last closed Chrome.

Like, you know when any software starts having stability issues it manifests always the same couple ways, like you could for example keep the arrow button pressed in a text edit field, and while the cursor rapidly paces in the text, it would occasionally lag and skip a few places. I don't even have that when the new tab with this extension basically already plays a slideshow instead of a 30 FPS video loop.

I close the browser, restart the browser, same tabs are open again (regardless how many, could be 200+), and the new tab is stable.

**Reply by Asagrim** [reply]
Is there an ETA on when this feature goes live?

**Reply by zombieFox** [reply]
No ETA on releasing this feature to Chrome and Firefox. Right now 7.0.2 is still rolling out to all user on those browsers. Once that is complete a new update will be published with this change. It can take days to a week or two for Chrome to review updates before publishing.

**Comment by zombieFox** [comment]
This feature is now out for Chrome and Firefox 🎸
